src/plugins/formulas/formulas.js

```diff
@@ -126,318 +166,562 @@ export class Formulas extends BasePlugin {
+   * @param {number} amount Number of newly created rows in the data source array.
+   * @returns {*|boolean} If false is returned the action is canceled.
    */
-  onBeforeCreateRow(row, amount, source) {
-    if (source === 'UndoRedo.undo') {
-      this.undoRedoSnapshot.restore();
+  onBeforeCreateRow(row, amount) {
+    if (!this.engine.isItPossibleToAddRows(this.sheetId, [this.hot.toPhysicalRow(row), amount])) {
+      return false;
     }
   }
 
   /**
-   * On after create row listener.
+   * `beforeCreateCol` hook callback.
    *
    * @private
-   * @param {number} row Row index.
-   * @param {number} amount An amount of created rows.
-   * @param {string} source Source of method call.
+   * @param {number} col Represents the visual index of first newly created column in the data source.
+   * @param {number} amount Number of newly created columns in the data source.
+   * @returns {*|boolean} If false is returned the action is canceled.
    */
-  onAfterCreateRow(row, amount, source) {
-    this.sheet.alterManager.triggerAlter('insert_row', row, amount, source !== 'UndoRedo.undo');
+  onBeforeCreateCol(col, amount) {
+    if (!this.engine.isItPossibleToAddColumns(this.sheetId, [this.hot.toPhysicalColumn(col), amount])) {
+      return false;
+    }
   }
 
   /**
-   * On before remove row listener.
+   * `beforeRemoveRow` hook callback.
    *
    * @private
-   * @param {number} row Row index.
-   * @param {number} amount An amount of removed rows.
+   * @param {number} row Visual index of starter row.
+   * @param {number} amount Amount of rows to be removed.
+   * @param {number[]} physicalRows An array of physical rows removed from the data source.
+   * @returns {*|boolean} If false is returned the action is canceled.
    */
-  onBeforeRemoveRow(row, amount) {
-    this.undoRedoSnapshot.save('row', row, amount);
+  onBeforeRemoveRow(row, amount, physicalRows) {
+    const possible = physicalRows.every((physicalRow) => {
+      return this.engine.isItPossibleToRemoveRows(this.sheetId, [physicalRow, 1]);
+    });
+
+    return possible === false ? false : void 0;
   }
 
   /**
-   * On after remove row listener.
+   * `beforeRemoveCol` hook callback.
    *
    * @private
-   * @param {number} row Row index.
-   * @param {number} amount An amount of removed rows.
+   * @param {number} col Visual index of starter column.
+   * @param {number} amount Amount of columns to be removed.
+   * @param {number[]} physicalColumns An array of physical columns removed from the data source.
+   * @returns {*|boolean} If false is returned the action is canceled.
    */
-  onAfterRemoveRow(row, amount) {
-    this.sheet.alterManager.triggerAlter('remove_row', row, amount);
+  onBeforeRemoveCol(col, amount, physicalColumns) {
+    const possible = physicalColumns.every((physicalColumn) => {
+      return this.engine.isItPossibleToRemoveColumns(this.sheetId, [physicalColumn, 1]);
+    });
```

Comment:
All `before` hooks should be a single call to the HF:
```suggestion
    const columnsToRemove = physicalColumns.map((physicalColumn) => [physicalColumn, 1]);
    const possible = this.engine.isItPossibleToRemoveColumns(this.sheetId, ...columnsToRemove);
```

Reply:
@wojciechczerniak Should it be done now, or later as a separate PR?